File: P4/ruby/Weapon.rb
# encoding: UTF-8

require_relative 'WeaponType'
require_relative 'WeaponToUI'

module Deepspace
	class Weapon
		def initialize (n, t, u) # Weapon(n : string, t : WeaponType, u : int)
			@name = n
			@type = t
			@uses = u
		end
		
		attr_reader :name
		attr_reader :type
		attr_reader :uses
		
		def self.newCopy (other)
			new(other.name, other.type, other.uses)
		end

		def getUIversion
			WeaponToUI.new(self)
		end
		
		def to_s
			getUIversion.to_s
		end

		def power
			@type.power
		end
		
		def useIt
			if @uses > 0
				@uses -= 1
				power
			else
				1.0
			end
		end
	end
end

if $0 == __FILE__
	puts Deepspace::Weapon.new(1, Deepspace::WeaponType::LASER, 3)
end

File: Java/src/deepspace/Weapon.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package deepspace;

public class Weapon implements CombatElement {
    private String name;
    private WeaponType type;
    private int uses;
    
    Weapon(String n, WeaponType w, int u) {
        name = n;
        type = w;
        uses = u;
    }
    
    Weapon(final Weapon other) {
        name = other.name;
        type = other.type;
        uses = other.uses;
    }
    
    public WeaponToUI getUIversion(){
        return new WeaponToUI(this);
    }
    
    @Override
    public String toString(){
        return getUIversion().toString();
    }
    
    WeaponType getType() {
        return type;
    }
    
    @Override
    public int getUses() {
        return uses;
    }
    
    public float power() {
        return type.getPower();
    }
    
    @Override
    public float useIt() {
        if(uses > 0){
            uses -= 1;
            return power();
        }else
            return 1;
    }
}